### PR TITLE
METRON-2250 Missing services in HDP 3.1 metron mpack and installer stuck

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

https://issues.apache.org/jira/browse/METRON-2250

Workaround for a bug in the Ambari. See ticket for more details.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- n/a Have you written or updated unit tests and or integration tests to verify your changes?
- n/a If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- n/a Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

- n/a Have you ensured that any documentation diagrams have been updated, along with their source files, using [draw.io](https://www.draw.io/)? See [Metron Development Guidelines](https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines) for instructions.

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by mmiklavc** [comment]
### Test Plan

In order reproduce the error being fixed in this bug, pull the latest feature branch code without this fix and follow the following test plan that also verifies the fix.

#### Testing in fulldev

1. Spin up full dev with Centos 7
    ```
    cd ./metron-deployment/development/centos7/
    vagrant up
    ```

#### Uninstall Metron service

1. Once the install completes, open Ambari http://node1:8080 and stop all Metron services under the Metron service option. We need to do this in order to uninstall.
2. Click Actions and choose delete service. Confirm.
3. The Metron Ambari service should be deleted

#### Re-install Metron service

1. Click the ellipses next to "Services" in Ambari and choose "Add Service"
2. Select Metron in the list of services to install. Accept the additional components by clicking "Ok" for each.
3. The "Assign Masters" page is where the bug previously existed. The "Next" option was greyed out due to the hostname errors. At this point, you can accept the defaults (node1 for full dev) and click "Next."
4. You can finish the installation if you like, but at this point the bug fix has been validated.

**Comment by mmiklavc** [comment]
@anandsubbu @MohanDV - would be great if you could take a look at this in a multi-node cluster and confirm it works as expected.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py [hunk 1]
```diff
@@ -41,21 +41,31 @@ def getServiceComponentLayoutValidations(self, services, hosts):
...
-        metronEnrichmentMaster = self.getHosts(componentsList, "METRON_ENRICHMENT_MASTER")[0]
-        metronProfilerHost = self.getHosts(componentsList, "METRON_PROFILER")[0]
-        metronPcapHost = self.getHosts(componentsList, "METRON_PCAP")[0]
-        metronIndexingHost = self.getHosts(componentsList, "METRON_INDEXING")[0]
-        metronRESTHost = self.getHosts(componentsList, "METRON_REST")[0]
-        metronManagementUIHost = self.getHosts(componentsList, "METRON_MANAGEMENT_UI")[0]
-        metronAlertsUIHost = self.getHosts(componentsList, "METRON_ALERTS_UI")[0]
+        stormSupervisors = self.getHosts(componentsList, "SUPERVISOR")
+
+        # This host assignment strategy for Metron hosts is an ugly hack to get around an unreported bug in Ambari 2.7.
```
Comment (justinleet) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py]: If it's unreported, we should report it and put the Jira number here instead.

**Comment by MohanDV** [comment]
@mmiklavc I have tested this on 5 node cluster.  This workaround will not solve the issue, this will just enable the 'NEXT'  button but it will not list the components, So this will cause the services unlisted after installation.  Also if you look at the `services.json` under `/var/run/ambari-server/stack-recommendations/#/`  It  has other components under different services like LIVY2_SERVER, SPARK2_THRIFTSERVER, PHOENIX_QUERY_SERVER, NFS_GATEWAY  also has empty host list. So this will cause problem in the while bringing up services.

**Comment by mmiklavc** [comment]
@MohanDV I'm a bit surprised a "recommendation" is being used for more than a recommendation. Sounds like it's used as part of the actual services startup.

**Comment by mmiklavc** [comment]
Closing as it looks like this will not be sufficient to fix the issue. Fixing this issue will require an Ambari code change.

**Pull request closed without merging**